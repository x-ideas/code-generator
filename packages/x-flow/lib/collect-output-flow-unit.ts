/*
 * @name:
 * description:
 *  用于收集所有子flow unit的output的数据，统一传递给下一个flow unit
 */

import { XGroupedFlowUnit } from './grouped-flow-unit';

export class CollectOutputFlowUnit extends XGroupedFlowUnit {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
  async doWork(data: any): Promise<any[]> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const result: any[] = [];

    let nextData = data;
    for (const unit of this._units) {
      nextData = await unit.doWork(nextData);
      result.push(nextData);
    }

    return result;
  }
}
