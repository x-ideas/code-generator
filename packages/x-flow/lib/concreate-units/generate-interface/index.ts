/* eslint-disable @typescript-eslint/ban-ts-comment */
/*
 * @name:
 * description:
 *  后台interface生成的处理逻辑
 */

import { InputData, JSONSchemaInput, quicktype, TypeScriptTargetLanguage } from 'quicktype-core';
import { JSONSchema4 } from 'json-schema';
import jscodeshift from 'jscodeshift';

import lodash from 'lodash';
import { XFlowUnit } from '../../flow-unit';
import { parserConfig } from '../../utils/jscodeshift-parser';

type TOption = {
  nicePropertyName: boolean;
  fInterfacePrefix?: string;
  bInterfacePrefix?: string;
};

/**
 * 将json scheme生成interface的处理流程
 * @export
 * @class BackEndInterfaceGenerateFlowUnit
 * @extends {XFlowUnit}
 */
export class InterfaceGenerateFlowUnit extends XFlowUnit {
  #options: TOption;

  constructor(options: TOption) {
    super();

    this.#options = options;
  }

  /**
   *
   * @param params topName不带IF,由nicePropertyName去添加
   */
  async doWork(params: { jsonSchema: JSONSchema4; topName: string }): Promise<string> {
    const inputData = new InputData();

    const source = { name: params.topName, schema: JSON.stringify(params.jsonSchema) };
    inputData.addSource('schema', source, () => new JSONSchemaInput(undefined));

    const { lines } = await quicktype({
      lang: new TypeScriptTargetLanguage(),
      inputData,

      // leadingComments: ['测试备注'],
      rendererOptions: {
        'just-types': 'true',
        // @ts-ignore
        'nice-property-names': this.#options.nicePropertyName ? 'true' : '',
        'acronym-style': 'camel',
      },
    });

    const prefix = this.#options.nicePropertyName ? this.#options.fInterfacePrefix ?? 'IF' : this.#options.bInterfacePrefix ?? 'IB';

    const replace1 = jscodeshift(lines.join('\n'), {
      parser: parserConfig(),
    })
      .find(jscodeshift.TSInterfaceDeclaration)
      .forEach(path => {
        //

        if (jscodeshift.Identifier.check(path.node.id)) {
          path.node.id.name = `${prefix}${path.node.id.name}`;
        }
      })
      .toSource();

    const replace2 = jscodeshift(replace1, {
      parser: parserConfig(),
    })
      .find(jscodeshift.TSTypeReference)
      .forEach(path => {
        //

        if (jscodeshift.Identifier.check(path.node.typeName)) {
          path.node.typeName.name = `${prefix}${path.node.typeName.name}`;
        }
      })
      .toSource();

    if (this.#options.nicePropertyName) {
      return replace2;
    } else {
      return replace2.replace(/export/g, '');
    }
  }
}
