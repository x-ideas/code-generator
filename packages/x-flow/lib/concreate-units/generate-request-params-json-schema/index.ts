/* eslint-disable @typescript-eslint/ban-ts-comment */
/*
 * @name:
 * description:
 *  解析swagger文档，得到request，并将其转换成json schema
 */

import { JSONSchema4 } from 'json-schema';
import { OpenAPIV2 } from 'openapi-types';
import lodash from 'lodash';

import { XFlowUnit } from '../../flow-unit';

interface IGenerateRequestParamsSchemaResult {
  path: JSONSchema4;
  body: JSONSchema4;
  query: JSONSchema4;
}

function isInBodyParameterObject(parameter: OpenAPIV2.Parameter): parameter is OpenAPIV2.InBodyParameterObject {
  return parameter.in === 'body';
}

export class GenerateRequestParamsJsonSchemaFlowUnit extends XFlowUnit {
  /**
   * 获取描述
   * @param obj
   */
  private getDescription(obj: OpenAPIV2.OperationObject) {
    if (obj.summary && obj.description) {
      return `${obj.summary} 
       ${obj.description}
      `;
    }
    return obj.summary ?? obj.description ?? '';
  }

  private _convertJsonTypeDefineFromOpenApi(parameter: OpenAPIV2.GeneralParameterObject): JSONSchema4 {
    switch (parameter.type) {
      case 'array':
        return {
          type: parameter.type,
          description: parameter.description,
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          items: this.generateItemJsonSchema(parameter.items, {}, []),
        };

      case 'object':
        return {
          type: parameter.type,
          description: parameter.description,
          properties: this.generatePropertiesJSONSchema(parameter.properties ?? {}, {}, []),
        };

      case 'string':
      case 'number':
      case 'integer':
      case 'boolean':
        return {
          type: parameter.type,
          description: parameter.description ?? '',
        };

      default:
        return {};
    }
  }

  /**
   * 解析path参数
   * @private
   * @param {OpenAPIV2.Parameter[]} parameters
   * @returns {{
   *     [name: string]: JSONSchema4;
   *   }}
   * @memberof GenerateRequestParamsJsonSchemaFlowUnit
   */
  private convertPathParemters(
    parameters: OpenAPIV2.Parameter[]
  ): {
    [name: string]: JSONSchema4;
  } {
    return (
      parameters
        .filter(par => {
          return par.in === 'path';
        })
        // @ts-ignore
        .map((par: OpenAPIV2.GeneralParameterObject) => {
          return {
            [par.name]: this._convertJsonTypeDefineFromOpenApi(par),
          };
        })
        .reduce((accum, current) => {
          return {
            ...accum,
            ...current,
          };
        }, {})
    );
  }

  /**
   * 获得必填的path参数
   * @private
   * @param {OpenAPIV2.Parameter[]} parameters
   * @returns {string[]}
   * @memberof GenerateRequestParamsJsonSchemaFlowUnit
   */
  private getRequeiredPathParemters(parameters: OpenAPIV2.Parameter[]): string[] {
    return parameters
      .filter(par => {
        return par.in === 'path';
      })
      .filter(par => {
        return par.required === true;
      })
      .map(par => par.name);
  }

  private generateItemJsonSchema(
    items?: OpenAPIV2.ItemsObject,
    defs: OpenAPIV2.DefinitionsObject = {},
    refs: string[] = []
  ): JSONSchema4 | JSONSchema4[] | undefined {
    if (items) {
      if (items.$ref) {
        if (!refs.includes(items.$ref)) {
          // @ts-ignore
          throw Error('此时不应该出现ref，马上联系管理员');
          // this.dealWithRef(items.$ref, defs, refs, items?.description);
        }
      }

      switch (items.type) {
        case 'array': {
          return {
            type: items.type,
            items: this.generateItemJsonSchema(items.items, defs, refs),
          };
        }

        case 'object': {
          // 此时ref被填充
          return {
            type: items.type,
            // @ts-ignore
            description: items.title ?? '',
            // @ts-ignore
            properties: this.generatePropertiesJSONSchema(items.properties ?? {}, defs, refs),
          };
        }

        case 'string':
        case 'number':
        case 'integer':
        case 'boolean':
          return {
            type: items.type,
            description: 'item-暂无',
          };
      }
    }

    return undefined;
  }

  private generateJSonSchemaFromOpenAPISchema(
    schema: OpenAPIV2.SchemaObject,
    defs: OpenAPIV2.DefinitionsObject,
    refs: string[],
    description?: string
  ): JSONSchema4 {
    if (schema.$ref && !refs.includes(schema.$ref)) {
      throw Error('此时不应该存在ref, 或者这是一个错误');
    }

    switch (schema.type) {
      case 'array': {
        const items = schema.items;
        return {
          type: schema.type,
          description: schema.description ?? schema.title ?? description,
          items: this.generateItemJsonSchema(items, defs, refs),
        };
      }

      case 'object': {
        return {
          type: schema.type,
          description: schema.description ?? description,
          properties: this.generatePropertiesJSONSchema(schema.properties ?? {}, defs, refs),
        };
      }

      case 'string':
      case 'number':
      case 'integer':
      case 'boolean':
        return {
          type: schema.type,
          description: schema.description ?? description,
          // default: schema.default,
          // required: schema.required,
        };

      default:
        return {};
    }
  }

  private generatePropertiesJSONSchema(
    properties: { [key: string]: OpenAPIV2.SchemaObject },
    defs: OpenAPIV2.DefinitionsObject,
    refs: string[],
    description = ''
  ): { [key: string]: JSONSchema4 } {
    const result: { [key: string]: JSONSchema4 } = {};
    for (const key of Object.keys(properties)) {
      const property = properties[key];

      result[key] = this.generateJSonSchemaFromOpenAPISchema(property, defs, refs, description);
    }

    return result;
  }

  private convertBodyParemeters(parameters: OpenAPIV2.Parameter[]): { [name: string]: JSONSchema4 } {
    return (
      parameters
        .filter(par => {
          return par.in === 'body';
        })
        // @ts-ignore
        .map((par: OpenAPIV2.GeneralParameterObject) => {
          if (par.schema) {
            // this.generateJSonSchemaFromOpenAPISchema(par.schema, {}, []);

            // const result: JSONSchema4 = {
            //   type: par.schema.type ?? 'object',
            //   description: par.schema.description ?? par.schema.title ?? '',
            //   properties: this.generatePropertiesJSONSchema(par.schema.properties ?? {}, {}, []),
            // };
            return {
              [par.name]: this.generateJSonSchemaFromOpenAPISchema(par.schema, {}, [], par.description),
            };
          }
          return {
            [par.name]: this._convertJsonTypeDefineFromOpenApi(par),
          };
        })
        .reduce((accum, current) => {
          return {
            ...accum,
            ...current,
          };
        }, {})
    );
  }

  private convertQueryParemeters(parameters: OpenAPIV2.Parameter[]): { [name: string]: JSONSchema4 } {
    return (
      parameters
        .filter(par => {
          return par.in === 'query';
        })
        // @ts-ignore
        .map((par: OpenAPIV2.GeneralParameterObject) => {
          return {
            [par.name]: this._convertJsonTypeDefineFromOpenApi(par),
          };
        })
        .reduce((accum, current) => {
          return {
            ...accum,
            ...current,
          };
        }, {})
    );
  }

  private getRequeiredQueryParemters(parameters: OpenAPIV2.Parameter[]): string[] {
    return parameters
      .filter(par => {
        return par.in === 'query';
      })
      .filter(par => {
        return par.required === true;
      })
      .map(par => par.name);
  }

  /**
   * 将swagger文档中的请求参数转换成符合scheme 结构的properties
   * @private
   * @param {OpenAPIV2.Parameter[]} parameters
   * @memberof GenerateRequestJsonSchema
   */
  private convertParameters(parameters: OpenAPIV2.Parameter[]): { [k: string]: JSONSchema4 } | undefined {
    const result = parameters
      .filter(par => {
        // 过滤掉header中的参数，因为是用户权限，统一设置
        return par.in !== 'header';
      })
      // .filter(par => {
      //   // 过滤掉path中的参数，这部分参数另外处理
      //   return par.in !== 'path';
      // })
      .map((par: OpenAPIV2.Parameter) => {
        if (par.in === 'path') {
          // 路径参数
          const result: JSONSchema4 = {
            path: {
              [par.name]: this._convertJsonTypeDefineFromOpenApi(par as OpenAPIV2.GeneralParameterObject),
            },
          };
          return result;
        }

        if (isInBodyParameterObject(par)) {
          const result: JSONSchema4 = {
            body: par.schema,
          };
          return result;
        } else {
          const result: JSONSchema4 = {
            [par.in]: {
              [par.name]: this._convertJsonTypeDefineFromOpenApi(par),
            },
          };
          return result;
        }
      })
      .reduce((accum, current) => {
        return lodash.merge(accum, current);
      }, {});

    return result;
  }

  private findRequestInPath(obj: OpenAPIV2.Document): OpenAPIV2.OperationObject | undefined {
    const paths: OpenAPIV2.PathItemObject[] = Object.values(obj.paths);
    if (paths[0]) {
      const firstPath = paths[0];

      return firstPath.get ?? firstPath.post ?? firstPath.delete ?? firstPath.put ?? undefined;
    }

    return undefined;
  }

  private getRequeiredParamers(parameters: OpenAPIV2.Parameter[]): string[] {
    return parameters
      .filter(par => {
        return par.in !== 'header' && par.in !== 'path';
      })
      .filter(par => {
        return par.required === true;
      })
      .map(par => {
        if (par.in === 'body') {
          return 'body';
        } else {
          return par.name;
        }
      });
  }

  async doWork(operationObj: OpenAPIV2.Document): Promise<IGenerateRequestParamsSchemaResult> {
    // 找到path中的请求对象

    const requestDef = this.findRequestInPath(operationObj);
    if (!requestDef) {
      throw Error('没有找到请求的定义');
    }

    return {
      path: {
        $schema: 'http://json-schema.org/draft-04/schema#',
        description: this.getDescription(requestDef),
        // NOTE: 以为先用swagger parse进行处理，我们这里先不考虑ref
        properties: this.convertPathParemters((requestDef.parameters ?? []) as OpenAPIV2.Parameter[]),
        required: this.getRequeiredPathParemters((requestDef.parameters ?? []) as OpenAPIV2.Parameter[]),
      },
      body: {
        $schema: 'http://json-schema.org/draft-04/schema#',
        description: this.getDescription(requestDef),
        properties: this.convertBodyParemeters((requestDef.parameters ?? []) as OpenAPIV2.Parameter[]),
        required: [],
      },
      query: {
        $schema: 'http://json-schema.org/draft-04/schema#',
        description: this.getDescription(requestDef),
        properties: this.convertQueryParemeters((requestDef.parameters ?? []) as OpenAPIV2.Parameter[]),
        required: this.getRequeiredQueryParemters((requestDef.parameters ?? []) as OpenAPIV2.Parameter[]),
      },
    };
  }
}
