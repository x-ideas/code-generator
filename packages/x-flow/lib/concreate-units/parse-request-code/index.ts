export * from './parser-request-code-unit';
