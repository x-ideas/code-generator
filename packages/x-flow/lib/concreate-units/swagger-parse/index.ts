export * from './swagger-parser-flow-unit';
