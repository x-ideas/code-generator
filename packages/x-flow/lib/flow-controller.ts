import { XFlowUnit } from './flow-unit';

type FuncType = (...args: unknown[]) => unknown;

function compose(funcs: FuncType[]): FuncType {
  return funcs.reduce((accum, current) => {
    return async function (data: unknown) {
      const nextData: any = await accum(data);
      return current(nextData);
    };
  });
}

export class XFlowController {
  private _units: XFlowUnit[];

  constructor() {
    this._units = [];
  }

  addUnit(unit: XFlowUnit): void {
    this._units.push(unit);
  }

  async run(data?: any): Promise<any> {
    const final = compose(this._units.map(unit => unit.doWork.bind(unit)));

    const result = final(data);
    return result;
  }
}
