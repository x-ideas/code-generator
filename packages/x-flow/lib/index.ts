export * from './concreate-units/generate-service-request';
