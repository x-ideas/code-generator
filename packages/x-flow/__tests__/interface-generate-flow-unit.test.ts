import { OpenAPIV2 } from 'openapi-types';
import { CollectOutputFlowUnit } from '../lib/collect-output-flow-unit';
import { InterfaceGenerateFlowUnit } from '../lib/concreate-units/generate-interface';
import { ParseRequestCodeFlowUnit } from '../lib/concreate-units/parse-request-code';
import { SpecialOutputFlowUnit } from '../lib/concreate-units/specify-output';
import { SwaggerParseFlowUnit } from '../lib/concreate-units/swagger-parse';
import { XFlowController } from '../lib/flow-controller';

import OpenAPIData from './assets/openAPI.json';
import { GenerateRequestParamsJsonSchemaFlowUnit } from '../lib/concreate-units/generate-request-params-json-schema';
import { GenerateResponseDataSchemaFlowUnit } from '../lib/concreate-units/generate-response-data-schema';

describe('测试InterfaceGenerateFlowUnit', () => {
  it('不修改property', async () => {
    const fc = new XFlowController();

    const collectOutputUnit = new CollectOutputFlowUnit();

    const outputCodeUnit = new SpecialOutputFlowUnit('3610401');
    const outputOpenAPIUnit = new SpecialOutputFlowUnit(OpenAPIData);
    collectOutputUnit.addUnit(outputCodeUnit);
    collectOutputUnit.addUnit(outputOpenAPIUnit);

    const codeParseUnit = new ParseRequestCodeFlowUnit();
    const swaggerParseUnit = new SwaggerParseFlowUnit();
    const frpUnit = new GenerateRequestParamsJsonSchemaFlowUnit();

    const backendInterfaceGenerate = new InterfaceGenerateFlowUnit({ nicePropertyName: false });

    fc.addUnit(collectOutputUnit);
    fc.addUnit(codeParseUnit);
    fc.addUnit(swaggerParseUnit);
    fc.addUnit(frpUnit);
    fc.addUnit(backendInterfaceGenerate);

    const result: OpenAPIV2.Document = await fc.run();
    // console.log(result);
  });

  test('3610401的响应-包含数组', async () => {
    const fc = new XFlowController();

    const collectOutputUnit = new CollectOutputFlowUnit();

    const outputCodeUnit = new SpecialOutputFlowUnit('3610401');
    const outputOpenAPIUnit = new SpecialOutputFlowUnit(OpenAPIData);
    collectOutputUnit.addUnit(outputCodeUnit);
    collectOutputUnit.addUnit(outputOpenAPIUnit);

    const codeParseUnit = new ParseRequestCodeFlowUnit();
    const swaggerParseUnit = new SwaggerParseFlowUnit();
    const frpUnit = new GenerateResponseDataSchemaFlowUnit();

    const backendInterfaceGenerate = new InterfaceGenerateFlowUnit({ nicePropertyName: false });

    fc.addUnit(collectOutputUnit);
    fc.addUnit(codeParseUnit);
    fc.addUnit(swaggerParseUnit);
    fc.addUnit(frpUnit);
    fc.addUnit(backendInterfaceGenerate);

    const result: OpenAPIV2.Document = await fc.run();
    // console.log(result);
  });

  test('3610116的响应-循环引用', async () => {
    const fc = new XFlowController();

    const collectOutputUnit = new CollectOutputFlowUnit();

    const outputCodeUnit = new SpecialOutputFlowUnit('3610116');
    const outputOpenAPIUnit = new SpecialOutputFlowUnit(OpenAPIData);
    collectOutputUnit.addUnit(outputCodeUnit);
    collectOutputUnit.addUnit(outputOpenAPIUnit);

    const codeParseUnit = new ParseRequestCodeFlowUnit();
    const swaggerParseUnit = new SwaggerParseFlowUnit();
    const frpUnit = new GenerateResponseDataSchemaFlowUnit();

    const backendInterfaceGenerate = new InterfaceGenerateFlowUnit({ nicePropertyName: false });

    fc.addUnit(collectOutputUnit);
    fc.addUnit(codeParseUnit);
    fc.addUnit(swaggerParseUnit);
    fc.addUnit(frpUnit);
    fc.addUnit(backendInterfaceGenerate);

    const result: OpenAPIV2.Document = await fc.run();
    // console.log(result);
  });
});
