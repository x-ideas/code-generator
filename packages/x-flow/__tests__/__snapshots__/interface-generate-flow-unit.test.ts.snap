// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试InterfaceGenerateFlowUnit--请求参数 只有header的请求 1`] = `""`;

exports[`测试InterfaceGenerateFlowUnit--请求参数 只有header的请求 2`] = `""`;

exports[`测试InterfaceGenerateFlowUnit--请求参数 只有header的请求 3`] = `""`;

exports[`测试InterfaceGenerateFlowUnit--请求参数 有header + query的请求 1`] = `""`;

exports[`测试InterfaceGenerateFlowUnit--请求参数 有header+path-id的请求 1`] = `
"export interface IFRegionInfoObject {
    /**
     * id
     */
    id: string;
}
"
`;

exports[`测试InterfaceGenerateFlowUnit--请求参数 有header+path-id的请求 2`] = `""`;

exports[`测试InterfaceGenerateFlowUnit--请求参数 有header+path-id的请求 3`] = `""`;

exports[`测试InterfaceGenerateFlowUnit--请求参数 有header+path-非id的请求 1`] = `
"export interface IFRegionInfoObject {
    /**
     * 门店补货需求单明细ID
     */
    detailId: string;
}
"
`;

exports[`测试InterfaceGenerateFlowUnit--请求参数 有header+path-非id的请求 2`] = `""`;

exports[`测试InterfaceGenerateFlowUnit--请求参数 有header+path-非id的请求 3`] = `""`;

exports[`测试InterfaceGenerateFlowUnit-response 3601401-data 1`] = `
"/**
 * 分页查找补货需求单: 3610401
 */
export interface IBDemo {
    count?: number;
    data?:  IBArray<IBDatum[]>;
    pager?: IBPager;
}

/**
 * 补货需求单，列表用
 */
export interface IBDatum {
    /**
     * 需求单号
     */
    apply_no?: string;
    /**
     * 补货类型,10=旧版采购补货,20=总仓自动补货,30=DC仓采购补货(手动)
     */
    apply_type?: number;
    /**
     * 计划到货时间段，10=白天，20=晚上
     */
    date_part_plan?: number;
    /**
     * 计划到货日期
     */
    date_plan?: number;
    /**
     * 总仓简称
     */
    dc_abbr_name?: string;
    /**
     * 总仓id
     */
    dc_id?: string;
    /**
     * 总仓名称
     */
    dc_name?: string;
    /**
     * id
     */
    id?: string;
    /**
     * 是否删除
     */
    is_delete?: boolean;
    /**
     * 计划品项数量
     */
    item_count_plan?: number;
    /**
     * 实际品项数量
     */
    item_count_real?: number;
    /**
     * 补货订单id
     */
    order_id?: string;
    /**
     * 状态，10=待发送，20=待发货，30=待收货，40=到货中，50=已完成
     */
    status?: number;
    /**
     * 门店简称
     */
    store_abbr_name?: string;
    /**
     * 门店code
     */
    store_code?: number;
    /**
     * 门店id
     */
    store_id?: string;
    /**
     * 门店名称
     */
    store_name?: string;
    /**
     * 创建时间
     */
    time_create?: number;
    /**
     * 计划到货时间
     */
    time_plan?: number;
    /**
     * 实际到货时间
     */
    time_real?: number;
    /**
     * 更新时间
     */
    time_update?: number;
    /**
     * 更新人id
     */
    user_id_update?: string;
    /**
     * 更新人名称
     */
    user_name_update?: string;
}

export interface IBPager {
    first_item_on_page?: number;
    has_next_page?:      boolean;
    has_previous_page?:  boolean;
    is_first_page?:      boolean;
    is_last_page?:       boolean;
    last_item_on_page?:  number;
    page_count?:         number;
    page_number?:        number;
    page_size?:          number;
    total_item_count?:   number;
}
"
`;

exports[`测试InterfaceGenerateFlowUnit-response 3610116的响应-循环引用 1`] = `
"export interface IFRegionIndo {
    applyCount?:       number;
    applyCountFinish?: number;
    /**
     * 补货类型：20=总仓自动补货，30=DC仓采购补货
     */
    applyType?:    number;
    batchCount?:   number;
    batchType?:    number;
    datePartPlan?: number;
    datePlan?:     number;
    dcId?:         string;
    dcName?:       string;
    hasDeleted?:   boolean;
    id?:           string;
    itemCount?:    number;
    num?:          string;
    /**
     * 概览信息
     */
    overview?: IFOverview;
    /**
     * 进度
     */
    progress?:       IFProgress;
    remark?:         string;
    status?:         number;
    timeCreate?:     number;
    timePlan?:       number;
    timeUpdate?:     number;
    userIdCreate?:   string;
    userIdUpdate?:   string;
    userNameCreate?: string;
    userNameUpdate?: string;
}

/**
 * 概览信息
 */
export interface IFOverview {
    /**
     * 基本信息
     */
    baseInfo?: IFBaseInfo;
    /**
     * 进度动态
     */
    process?: IFProcess;
}

/**
 * 基本信息
 */
export interface IFBaseInfo {
    /**
     * 集货数
     */
    collectCount?: number;
    /**
     * 品项数
     */
    itemCount?: number;
    /**
     * 波次数
     */
    orderCount?: number;
    /**
     * 波次列表
     */
    orderList?: any[];
    /**
     * 门店数
     */
    storeCount?: number;
    /**
     * 小时
     */
    timeGroup?: number;
    /**
     * 体积(m³)
     */
    volume?: number;
    /**
     * 重量(g)
     */
    weight?: number;
}

/**
 * 进度动态
 */
export interface IFProcess {
    /**
     * 集货进度
     */
    collectProgress?: number;
    /**
     * 相对集货进度
     */
    collectProgressRelative?: number;
    /**
     * 校正集货数总和
     */
    collectSumCorrect?: number;
    /**
     * 待处理集货数总和
     */
    collectSumPending?: number;
    /**
     * 计划集货数总和
     */
    collectSumPlan?: number;
    /**
     * 实际集货数总和
     */
    collectSumReal?: number;
    /**
     * 出货进度
     */
    deliveryProgress?: number;
    /**
     * 相对出货进度
     */
    deliveryProgressRelative?: number;
    /**
     * 实际出货数总和
     */
    deliverySumReal?: number;
    /**
     * 分货进度
     */
    distributeProgress?: number;
    /**
     * 相对分货进度
     */
    distributeProgressRelative?: number;
    /**
     * 实际分货数总和
     */
    distributeSumReal?: number;
    /**
     * 波次创建进度
     */
    orderProgress?: number;
    /**
     * 收货进度
     */
    receiveProgress?: number;
    /**
     * 相对收货进度
     */
    receiveProgressRelative?: number;
    /**
     * 实际入库数总和
     */
    receiveSumReal?: number;
}

/**
 * 进度
 */
export interface IFProgress {
    /**
     * 集货进度
     */
    collectProgress?: number;
    /**
     * 相对集货进度
     */
    collectProgressRelative?: number;
    /**
     * 校正集货数总和
     */
    collectSumCorrect?: number;
    /**
     * 待处理集货数总和
     */
    collectSumPending?: number;
    /**
     * 计划集货数总和
     */
    collectSumPlan?: number;
    /**
     * 实际集货数总和
     */
    collectSumReal?: number;
    /**
     * 出货进度
     */
    deliveryProgress?: number;
    /**
     * 相对出货进度
     */
    deliveryProgressRelative?: number;
    /**
     * 实际出货数总和
     */
    deliverySumReal?: number;
    /**
     * 分货进度
     */
    distributeProgress?: number;
    /**
     * 相对分货进度
     */
    distributeProgressRelative?: number;
    /**
     * 实际分货数总和
     */
    distributeSumReal?: number;
    /**
     * 波次创建进度
     */
    orderProgress?: number;
    /**
     * 收货进度
     */
    receiveProgress?: number;
    /**
     * 相对收货进度
     */
    receiveProgressRelative?: number;
    /**
     * 实际入库数总和
     */
    receiveSumReal?: number;
}
"
`;

exports[`测试InterfaceGenerateFlowUnit-response 3610401的响应-包含数组 1`] = `
"/**
 * 补货需求单，列表用
 */
export interface IBRegionIndo {
    /**
     * 需求单号
     */
    apply_no?: string;
    /**
     * 补货类型,10=旧版采购补货,20=总仓自动补货,30=DC仓采购补货(手动)
     */
    apply_type?: number;
    /**
     * 计划到货时间段，10=白天，20=晚上
     */
    date_part_plan?: number;
    /**
     * 计划到货日期
     */
    date_plan?: number;
    /**
     * 总仓简称
     */
    dc_abbr_name?: string;
    /**
     * 总仓id
     */
    dc_id?: string;
    /**
     * 总仓名称
     */
    dc_name?: string;
    /**
     * id
     */
    id?: string;
    /**
     * 是否删除
     */
    is_delete?: boolean;
    /**
     * 计划品项数量
     */
    item_count_plan?: number;
    /**
     * 实际品项数量
     */
    item_count_real?: number;
    /**
     * 补货订单id
     */
    order_id?: string;
    /**
     * 状态，10=待发送，20=待发货，30=待收货，40=到货中，50=已完成
     */
    status?: number;
    /**
     * 门店简称
     */
    store_abbr_name?: string;
    /**
     * 门店code
     */
    store_code?: number;
    /**
     * 门店id
     */
    store_id?: string;
    /**
     * 门店名称
     */
    store_name?: string;
    /**
     * 创建时间
     */
    time_create?: number;
    /**
     * 计划到货时间
     */
    time_plan?: number;
    /**
     * 实际到货时间
     */
    time_real?: number;
    /**
     * 更新时间
     */
    time_update?: number;
    /**
     * 更新人id
     */
    user_id_update?: string;
    /**
     * 更新人名称
     */
    user_name_update?: string;
}
"
`;

exports[`测试InterfaceGenerateFlowUnit-response 不修改property 1`] = `
"/**
 * 补货需求单，列表用
 */
export interface IBRegionIndo {
    /**
     * 需求单号
     */
    apply_no?: string;
    /**
     * 补货类型,10=旧版采购补货,20=总仓自动补货,30=DC仓采购补货(手动)
     */
    apply_type?: number;
    /**
     * 计划到货时间段，10=白天，20=晚上
     */
    date_part_plan?: number;
    /**
     * 计划到货日期
     */
    date_plan?: number;
    /**
     * 总仓简称
     */
    dc_abbr_name?: string;
    /**
     * 总仓id
     */
    dc_id?: string;
    /**
     * 总仓名称
     */
    dc_name?: string;
    /**
     * id
     */
    id?: string;
    /**
     * 是否删除
     */
    is_delete?: boolean;
    /**
     * 计划品项数量
     */
    item_count_plan?: number;
    /**
     * 实际品项数量
     */
    item_count_real?: number;
    /**
     * 补货订单id
     */
    order_id?: string;
    /**
     * 状态，10=待发送，20=待发货，30=待收货，40=到货中，50=已完成
     */
    status?: number;
    /**
     * 门店简称
     */
    store_abbr_name?: string;
    /**
     * 门店code
     */
    store_code?: number;
    /**
     * 门店id
     */
    store_id?: string;
    /**
     * 门店名称
     */
    store_name?: string;
    /**
     * 创建时间
     */
    time_create?: number;
    /**
     * 计划到货时间
     */
    time_plan?: number;
    /**
     * 实际到货时间
     */
    time_real?: number;
    /**
     * 更新时间
     */
    time_update?: number;
    /**
     * 更新人id
     */
    user_id_update?: string;
    /**
     * 更新人名称
     */
    user_name_update?: string;
}
"
`;
